### PR TITLE
Add contract interaction to preview dialog

### LINKED ISSUE #3731: Add "Contract interaction" transaction type to preview dialog
### Story

As a user creating a "Contract interaction" transaction from the Gnosis control dialog, I want to see a preview of the transaction before sending it.

### Description

When the user chooses the "Contract interaction", and fills the form of the Gnosis control dialog, after confirming the form and before sending the transaction, they should see a preview of what transactions they are going to send, along with the preview of other transactions (in the case of the user doing a multi transaction).

### Design

[Figma link](https://www.figma.com/file/LlVmeMQxHVA04evfrgZyuN/Safe-Control?node-id=5418%3A22229)
FireShot Capture 687 - Safe Control – Figma - www figma com

### This issue is complete when...

- [ ] You click on the submit button of the previous Gnosis Safe Control dialog screen with a "Contract interaction" transaction added and it takes you to the preview with a summary of that transaction.
- [ ] The summary includes the fields that could be parsed from the parameters of the function selected for the interaction.

### PR DESCRIPTION
**Pull request by Julianahlk**
## Description

This PR adds the Contract Interaction preview to the Safe form.

[Figma Design](https://www.figma.com/file/LlVmeMQxHVA04evfrgZyuN/Safe-Control?node-id=115%3A7477)

**To test:**

* Add a safe
* Click Contract Interaction
* Add a verified contract address (find a verified contracts here: https://blockscout.com/xdai/mainnet/api-docs#contract or https://etherscan.io/contractsVerified)
* Select a function to interact with - you can play around with different functions and make sure all their data is showing up in the preview correctly
* Test with one or multiple transactions - Make sure they are all in the preview

Changes 🏗

* Adding Contract Interaction preview

Screen Shot 2022-09-21 at 10 47 11 PM

Screen Shot 2022-09-21 at 11 30 21 PM

---
Resolves #3731

### CODE REVIEW

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx
Comment (ArmandoGraterol) [inline src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx:25]: This should be up

**Review by ArmandoGraterol — APPROVED** [review]
Nice job! Tested various functions with different data types and they are displaying nicely for me :+1:

**Comment by chinins** [comment]
> In #3891, I add to the validation so it validates arrays correctly. This is something to consider in this issue, how to display array data.
> 
> This is how it would look currently: sp
> 
> This was how I thought it might look: nosp
> 
> So, opening that up for discussion. Which do you think looks better? Spaces or no?
> 
> Also, in this specific example, I presume you'd display an array of addresses with the same shortening that you apply to single addresses?

I would go for the first option. Maybe doesn't look as fancy but at least you can identify and check addresses. In the second option it's just a mess of info.

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx
Comment (Julianahlk) [reply src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx]: Done! :)

**Comment by Julianahlk** [comment]
> Nice job! Tested various functions with different data types and they are displaying nicely for me 👍

Thanks Armando! Already made the import change 👍.

**Review by chinins — APPROVED** [review]
Looks good!

**Comment by Julianahlk** [comment]
> So, opening that up for discussion. Which do you think looks better? Spaces or no?
> 
> Also, in this specific example, I presume you'd display an array of addresses with the same shortening that you apply to single addresses?

@willm30, thanks for bringing the arrays validation case to my attention. From both suggested options above, I agree with @chinins about the first option being the best for this case because it is definitely easier to read and identify where the address starts and ends.

**Comment by Julianahlk** [comment]
@willm30 After doing some more investigation it seems that arrays appear like the first option you suggested when there is a space after the comma between addresses and the 2nd option when there is no space after the comma. Are you suggesting that I alter the address array string input to make it appear like the first option? by parsing the array string and adding a space after the commas? 

Besides, would a user ever be able to manually input any array data? 

I feel like we could wait until you merge your [PR](https://github.com/JoinColony/colonyDapp/pull/3891) that allows the input of an array of addresses as it may be more clear what the ideal behavior would be. 

Let me know your thoughts. Thanks!

**Comment by willm30** [comment]
@Julianahlk 

> Besides, would a user ever be able to manually input any array data?

Yes, this will be possible.  

> I feel like we could wait until you merge your [PR]

Sounds good. You could rebase on it now or wait until it's been merged. Validation doesn't enforce spaces after commas in arrays, so if you need that to get the styling to work correctly, you could call [getArrayFromString](https://github.com/JoinColony/colonyDapp/blob/0fbcdc40a5e3ca49df1f39b85849573238bfb3f3/src/utils/safes/contractParserValidation.ts#L61) and then `.join(',  ')` and put the brackets back on.

Or display it differently. Could be an indexed list like:

0: xxx
1: yyy  
2: zzz

Only a suggestion. Whatever looks best.

**Comment by willm30** [comment]
@Julianahlk I have merged #3891, FYI

**Comment by Julianahlk** [comment]
> @Julianahlk
> 
> > Besides, would a user ever be able to manually input any array data?
> 
> Yes, this will be possible.
> 
> > I feel like we could wait until you merge your [PR]
> 
> Sounds good. You could rebase on it now or wait until it's been merged. Validation doesn't enforce spaces after commas in arrays, so if you need that to get the styling to work correctly, you could call [getArrayFromString](https://github.com/JoinColony/colonyDapp/blob/0fbcdc40a5e3ca49df1f39b85849573238bfb3f3/src/utils/safes/contractParserValidation.ts#L61) and then `.join(', ')` and put the brackets back on.
> 
> Or display it differently. Could be an indexed list like:
> 
> 0: xxx 1: yyy 2: zzz
> 
> Only a suggestion. Whatever looks best.

@willm30 I am trying to update this as discussed, I rebased so I have the changes from your PR #3891 but I am not able to get past the control safe form because the create transaction button remains disabled even after filling in the information for all the input fields. There is no error message so I am not sure what is the problem:

error

**Comment by willm30** [comment]
@Julianahlk As discussed earlier, this was due to using a safe added prior to the ui update (which did not have a moduleContractAddress). Now fixed.

**Comment by Julianahlk** [comment]
> @Julianahlk As discussed earlier, this was due to using a safe added prior to the ui update (which did not have a moduleContractAddress). Now fixed.

@willm30 made the requested changes. Please take a look again whenever you can. Thank you!

Screen Shot 2022-09-30 at 10 02 58 PM

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx:280]: What about all other array types?

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx:277]: Why not an array of masked addresses?

**Review by willm30 — COMMENTED** [review]
Couple of comments in the code. Github actions is flagging a type error in `initiateSafeTransaction`.

Also noticed that we don't have the contract name per [this screen](https://www.figma.com/file/LlVmeMQxHVA04evfrgZyuN/Safe-Control?node-id=5416%3A22878), so have opened [an issue](https://github.com/JoinColony/colonyDapp/issues/3932) for that.

**Comment by Julianahlk** [comment]
@willm30 All types of arrays should be displaying correctly now, please review again whenever you get a chance. 

Screen Shot 2022-10-04 at 12 09 31 PM
The two Rinkeby contract addresses below contain different examples of arrays:
```
0x44762835EeB462fCe0150935593Fd119A935B16a
Function name: unstake
```

```
0xE65841eb30548E7be3FFB9e496DB1B540C5eD8C6
Function name: addLeaderboardEntry

```

I also, noticed in the feature branch the validation is not working correctly for multidimensional arrays:
Screen Shot 2022-10-04 at 12 12 48 PM

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx:278]: Consider passing `address={address.trim()}` here to avoid "Address format is wrong!":

address-wrong

**Review by willm30 — APPROVED** [review]
Looking good! Just a couple of comments.

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx
Comment (Julianahlk) [reply src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx:278]: Sounds good!

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx
Comment (Julianahlk) [reply src/modules/dashboard/components/Dialogs/ControlSafeDialog/SafeTransactionPreview.tsx:289]: Yes! Great idea :) Thanks Will.

**Pull request merged**
